File: problem_set/problem_set_2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    string text = get_string("Text: ");
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // L is the average number of letters per 100 words in the text
    float L = (float) letters / words * 100;

    // S is the average number of sentences per 100 words in the text.
    float S = (float) sentences / words * 100;

    // Coleman-Liau index
    int grade = round(0.0588 * L - 0.296 * S - 15.8);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

int count_letters(string text)
{
    int n = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (islower(text[i]) || isupper(text[i]))
        {
            n++;
        }
    }

    return n;
}

int count_words(string text)
{
    int n = 0;
    int isword = 0;

    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == ' ')
        {
            isword = 0;
        }
        else
        {
            if (isword == 0)
            {
                n++;
            }
            isword = 1;
        }
    }

    return n;
}

int count_sentences(string text)
{
    int n = 0;

    // Don't consider abbreviations, like "Mr. Robot?"
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            n++;
        }
    }

    return n;
}

File: problem_set/problem_set_6/Readability/readability.py
from cs50 import get_string
import re


def main():
    text = get_string("Text: ")
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # L is the average number of letters per 100 words in the text
    L = letters / words * 100

    # S is the average number of sentences per 100 words in the text.
    S = sentences / words * 100

    # Coleman-Liau index
    grade = round(0.0588 * L - 0.296 * S - 15.8)

    if grade < 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade}")


def count_letters(text):
    n = 0
    for char in text:
        match = re.match("^[a-zA-Z]$", char)
        if match:
            n += 1
    return n


def count_words(text):
    n = 0
    is_word = False

    for char in text:
        if char == " ":
            is_word = False
        elif is_word == False:
            n += 1
            is_word = True
    return n


def count_sentences(text):
    n = 0
    # Don't consider abbreviations, like "Mr. Robot?"
    for char in text:
        if char in [".", "!", "?"]:
            n += 1
    return n


main()
